keras_maskrcnn/losses.py

```diff
@@ -28,60 +28,91 @@ def _mask(y_true, y_pred, iou_threshold=0.5, mask_size=(28, 28)):
-        boxes                = keras_retinanet.backend.gather_nd(boxes, indices)
-        masks                = keras_retinanet.backend.gather_nd(masks, indices)
-        argmax_overlaps_inds = keras.backend.cast(keras_retinanet.backend.gather_nd(argmax_overlaps_inds, indices), 'int32')
-        labels               = keras.backend.cast(keras.backend.gather(annotations[:, 4], argmax_overlaps_inds), 'int32')
-
-        # make normalized boxes
-        x1 = boxes[:, 0]
-        y1 = boxes[:, 1]
-        x2 = boxes[:, 2]
-        y2 = boxes[:, 3]
-        boxes = keras.backend.stack([
-            y1 / (keras.backend.cast(height, dtype=keras.backend.floatx()) - 1),
-            x1 / (keras.backend.cast(width, dtype=keras.backend.floatx()) - 1),
-            (y2 - 1) / (keras.backend.cast(height, dtype=keras.backend.floatx()) - 1),
-            (x2 - 1) / (keras.backend.cast(width, dtype=keras.backend.floatx()) - 1),
-        ], axis=1)
-
-        # crop and resize masks_target
-        masks_target = keras.backend.expand_dims(masks_target, axis=3)  # append a fake channel dimension
-        masks_target = backend.crop_and_resize(
-            masks_target,
-            boxes,
-            argmax_overlaps_inds,
-            mask_size
-        )
-        masks_target = masks_target[:, :, :, 0]  # remove fake channel dimension
+        def _mask(args):
+            boxes = args[0]
+            masks = args[1]
+            annotations = args[2]
+            masks_target = args[3]
 
-        # gather the predicted masks using the annotation label
-        masks = backend.transpose(masks, (0, 3, 1, 2))
-        label_indices = keras.backend.stack([
-            keras.backend.arange(keras.backend.shape(labels)[0]),
-            labels
-        ], axis=1)
-        masks = keras_retinanet.backend.gather_nd(masks, label_indices)
+            return compute_loss(
+                boxes,
+                masks,
+                annotations,
+                masks_target,
+                width,
+                height,
+                iou_threshold = iou_threshold,
+                mask_size     = mask_size,
+            )
 
-        # compute mask loss
-        mask_loss  = keras.backend.binary_crossentropy(masks_target, masks)
-        normalizer = keras.backend.shape(masks)[0] * keras.backend.shape(masks)[1] * keras.backend.shape(masks)[2]
-        normalizer = keras.backend.maximum(keras.backend.cast(normalizer, keras.backend.floatx()), 1)
-        mask_loss  = keras.backend.sum(mask_loss) / normalizer
+        mask_batch_loss = keras_retinanet.backend.map_fn(
+            _mask,
+            elems=[boxes, masks, annotations, masks_target],
+            dtype=keras.backend.floatx(),
+            parallel_iterations=parallel_iterations
+        )
 
-        return mask_loss
+        return keras.backend.mean(mask_batch_loss)
 
     return _mask_conditional
+
+
+def compute_loss(
```

Comment:
Since this is a toplevel function, I suggest to rename it to `compute_mask_loss`. I had half a mind to move this function inside the scope of `mask`..